### PR TITLE
Reintroduce removal of stale *.pyc files #3895

### LINKED ISSUE #3895: Remove *pyc files
This will make you smile @jamalex :)

In #3783, I removed django_extensions which was just being used by the update command (removed) and to run the management command `clean_pyc`. This made the codebase over 33,000 lines shorter :)

The `clean_pyc` command is pretty much equivalent to `find -name '*pyc' -exec rm {} \;` or something.

So we need a simple python-version of this command to run in the "setup" management command to make up for its removal.

### PR DESCRIPTION
**Pull request by benjaoming**
Fixes #3895

Example:

```
➜  ka-lite git:(clean_pyc) ✗ bin/kalite manage setup   
                                  
  _   __  ___    _     _ _        
 | | / / / _ \  | |   (_) |       
 | |/ / / /_\ \ | |    _| |_ ___  
 |    \ |  _  | | |   | | __/ _ \ 
 | |\  \| | | | | |___| | ||  __/ 
 \_| \_/\_| |_/ \_____/_|\__\___| 
                                  
http://kalite.learningequality.org
                                  
         version 0.14.dev11
                                  
WARNING:root:It's recommended that you install Python version 2.7.9. Your version is: sys.version_info(major=2, minor=7, micro=6, releaselevel='final', serial=0)

--------------------------------------------------------------------------------
This script will configure the database and prepare it for use.
--------------------------------------------------------------------------------
Press [enter] to continue...
-------------------------------------------------------------------
WARNING: Database file already exists!
-------------------------------------------------------------------
Keep database file and upgrade to KA Lite version 0.14.dev11? (yes or no) yes
Upgrading database to KA Lite version 0.14.dev11
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/kalite/settings.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/kalite/knowledgemap/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/kalite/distributed/templatetags/kalite_staticfiles.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/kalite/distributed/templatetags/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/six.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/mock.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/base.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/fun.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/exc.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/util.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/typ.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/pack.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/stream.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/base.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/mem.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/git.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/loose.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/pack.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/gitdb/db/ref.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/django_extensions/models.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/django_extensions/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/models.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/logger.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/utils/py3.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/utils/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/db/sqlite3.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/db/generic.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/south/db/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/case.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/main.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/util.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/runner.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/suite.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/loader.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/compatibility.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/collector.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/result.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/unittest2/signals.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/config.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/exc.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/odict.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/diff.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/util.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/db.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/remote.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/cmd.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/repo/base.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/repo/fun.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/repo/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/index/base.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/index/fun.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/index/util.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/index/typ.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/index/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/reference.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/log.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/remote.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/tag.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/head.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/refs/symbolic.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/base.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/fun.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/util.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/tree.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/blob.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/commit.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/tag.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/submodule/base.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/submodule/util.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/submodule/root.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/git/objects/submodule/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/pytz/tzinfo.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/pytz/exceptions.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/pytz/tzfile.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/pytz/__init__.pyc
Removing old PYC file: /google_and_microsoft/shared_main_drive/very_interesting_projects/ka-lite/python-packages/fle_utils/testing/unicode.pyc
Creating tables ...
Installing custom SQL ...
Installing indexes ...
Installed 0 object(s) from 0 fixture(s)
CommandError: Unknown command: 'migrate'
```

### CODE REVIEW

**Comment by benjaoming** [comment]
Tested with a non-writable source dir, it just throws a warning. I think that's acceptable given that someone might be running on an okay but still non-writable source dir.

**Comment by benjaoming** [comment]
`settings.SOURCE_DIR` = ka-lite is checked out or downloaded directly from github

**Comment by rtibbles** [comment]
:+1: for this!

Also, N.B. it's ```[ci skip]``` :)

**Comment by benjaoming** [comment]
@rtibbles wow, they changed it? Thanks for the tip!! I was wondering why it didn't work but too lazy to investigate :D

**Comment by rtibbles** [comment]
Wondering if this is worth abstracting out, and running during kaserve if DEBUG=True too?

**Comment by benjaoming** [comment]
@rtibbles the thing is this:

If you use python-packages with pip or git, it's managed as nothing more than data. If a file is removed, a stale *pyc is left, even if setuptools/pip has done an update (because a stale *pyc is non-managed).

So in case we have done file removals in python-packages (as I've been doing a lot lately), it becomes relevant.

I should still actually look at python-packages for the case of pip-managed installations, because in the future, removing files in python-packages will continue to break things :(

**Comment by rtibbles** [comment]
Yep, no argument from me there, @benjaoming - I was just wondering if in addition to what you have here during setup, we could do something similar in kaserve when DEBUG=True, so that devs get the same utility during development.

**Comment by benjaoming** [comment]
@rtibbles I was actually quite happy about removing it because it was causing the dev server to run slower :) Wouldn't you think it's nicer to not have to scan all source files every time the dev server runs?

It won't fix everything regarding stale *pyc files anyways.. if you remove files while running `kaserve` in debug, it will reload current sources, but it won't re-run the management command. That means that devs would still be forced to CTRL+C the server and restart it or `kalite stop` and `kalite start` if they are removing source files or have done a `git pull` while kalite was active.

**Comment by rtibbles** [comment]
Fair enough - would there be any value add in having it as an additional management command (over and above clean_pyc), so that our less bash experienced devs could run it if needed? Running ```find -name "*.pyc" -delete``` doesn't work in Windows (naturally) nor directly in Git bash on Windows, so it might help some of our dev interns.

**Comment by benjaoming** [comment]
They will have to learn ;)

But re-running `kalite manage setup` is a way to achieve this.

**Comment by rtibbles** [comment]
Righto, let's merge this then.

**Comment by MCGallaspy** [comment]
@rtibbles in Windows, you can use the `del` statement to delete patterns, and also there's a recursive flag, so essentially the same command exists.

File: kalite/distributed/management/commands/setup.py [hunk 1]
```diff
@@ -263,7 +264,20 @@ def handle(self, *args, **options):
...
+        # Clean *pyc files if we are in a git repo
+        if not install_clean and settings.IS_SOURCE:
+            if os.access(settings.SOURCE_DIR, os.W_OK):
+                for root, dirs, files in os.walk(settings.SOURCE_DIR):
+                    pyc_files = filter(lambda filename: filename.endswith(".pyc"), files)
+                    py_files = set(filter(lambda filename: filename.endswith(".py"), files))
+                    excess_pyc_files = filter(lambda pyc_filename: pyc_filename[:-1] not in py_files, pyc_files)
+                    for excess_pyc_file in excess_pyc_files:
+                        full_path = os.path.join(root, excess_pyc_file)
+                        os.remove(full_path)
```
Comment (jamalex) [inline kalite/distributed/management/commands/setup.py]: Perhaps we should set a flag here if any files did get removed, and then restart the process after, if possible?

Comment (jamalex) [inline kalite/distributed/management/commands/setup.py]: (I still got the "CommandError: Unknown command: 'migrate'" error while running setup)

**Pull request merged**